src/detail/task_queue.cpp

```diff
@@ -48,10 +48,10 @@ task_queue::task_queue()
                   // NOTE: wait() is noexcept.
                   this->m_cond.wait(lock, [this]() { return this->m_stop || !this->m_tasks.empty(); });
 
-                  if (this->m_stop && this->m_tasks.empty()) {
-                      // If the stop flag was set, and we do not have more tasks,
-                      // just exit.
-                      break;
+                  if (this->m_tasks.empty()) {
+                      // If we do not have more tasks check stop and park flags
+                      if (this->m_stop) break;
+                      if (this->m_park) m_parked.notify_all();
```

Comment:
Not sure if I am parsing this correctly, but what happens after this ``notify_all()``? Are we ending up popping a task from the queue below even if the queue is empty?

Reply:
We're communicating about 2 different events in 2 different directions, so by my understanding we can either use 1 condition variable and use `notify_all` everywhere (and accept some extra spurious wakeups), or we could use 2 condition variables (one for each type of event) and use `notify_one` safe in the knowledge that there should only be a single waiting thread. Either way I agree the inconsistency at present could do with being cleaned up.

Reply:
The notify on `m_parked` is to signal to a waiting thread that the `task_queue` has completed all work on its thread following a call to `park`. This is to emulate the waiting behaviour of the `stop()` method (which is done by a call to `thread.join()` which we can't do here)  in the `park()` method.